### ISSUE TITLE
Add the ability to remove certs and keys

### ISSUE THREAD

**Issue by ameneau** [saltstack-formulas/cert-formula#15]
This formula doesn't offer a way to remove certificates and keys. I'll have a pull request to add in this functionality, however I was curious if there was any preference for doing this. I can do it one of two ways.

1. Create an "absent" state, put the contents of init.sls into a "present" state and have init.sls include the "absent" state, then the "present" state
2. Same as 1, except replacing init.sls with a "managed" state.

I would go with option 2, personally

**Comment by javierbertoli** [comment]
I don't fully understand your 2nd. option, but how about

1. Adding a new parameter to certs, like

```yaml
  certlist:
    name_of_cert:
    present: true        # (default to true if not set)
      key: |
KEY HERE
```
2. Then, in `init.sls`, do something like
2.1. [Here](https://github.com/saltstack-formulas/cert-formula/blob/master/cert/init.sls#L17)

```yaml
{% set state = 'managed' if data.get('present') is set else 'absent' %}
```
2.2. then, in the [`file.managed` lines like this one](https://github.com/saltstack-formulas/cert-formula/blob/master/cert/init.sls#L31) replace them by:

```yaml
  file.{{ state }}:
```

Of course, haven't even tested this, the `present` var can be replaced by any other meaningful name you can think of (like, `enabled` or... whatever) and probably, all the other vars should be automatically assigned default values so the state does not fail. Just an idea, based on your comment :smile:

I think it would make it easier to take advantage of the existing code, wdyt?

**Comment by ameneau** [comment]
Thanks for your quick reply and I apologize for the lack of clarity! I did not supply enough detail, sorry about that!

in option 1, you'd have the following three files in the "cert" folder
* init.sls
* present.sls
* absent.sls

In option 2, instead you'd have this instead:
* managed.sls
* present.sls
* absent.sls

These would then be separate states that you could apply in your top file or by executing a state.apply (ex cert.managed, cert.present, cert.absent).

I was originally planning to not touch the certlist pillar data and add another set of pillar data for the case where you want to have the certs or keys removed:

```
  certlist:
    name_of_cert:
      key: |
        KEY HERE
  absent:
    name_of_cert_to_remove:
      key: |
        KEY HERE
      cert: |
        CERT HERE
```

Adding a key/value pair of "present" to the certlist as you had mentioned has the advantage of not allowing for problematic data duplication (certlist has cert named "xyz" and absent also has cert named "xyz", what happens then? do you remove the cert and then add it back?). However, it is also nice to have that visual separation in the pillar file between certs/keys you want to absent and those that you want present.

Either way makes sense to me, if you have a preference for how the pillar data should look, I'm happy to conform to that.

As for replacing the file.managed line with file.{{ state }}, I think it'll be clearer to have the cert removal be a separate state (I think absent.sls makes sense), and I'm not sure file.{{ state }} would work well anyway since file.absent and file.managed likely have different options.

Thanks again, and hopefully this paints a clearer picture of the modifications I was planning to make. I'm also happy to submit a PR, get feedback and make additional changes based on the feedback.

**Comment by javierbertoli** [comment]
The rationale for my proposal was, as you mention, to avoid data duplication in pillars, in the first place: just adding a parameter (with a default of 'adding the cert' if not specified). This way, the formula would behave all the same as it does currently and the change will be backward compatible.

As you point out, `managed` and `absent` have different parameters, my example was more 'pseudocode'. 

You can manage it like [this](https://github.com/saltstack-formulas/rspamd-formula/blob/master/rspamd/config.sls#L16-L25) and [this](https://github.com/saltstack-formulas/fail2ban-formula/blob/master/fail2ban/config.sls#L7-L16) which are actual examples of current formulas.

My idea is that the pillar, in the end, would look like

```yaml
cert:
  certlist:
    name_of_cert:
      key: |
        KEY HERE
    name_of_2nd_cert:
      cert: |
        CERT HERE
      key: |
        KEY HERE
    name_of_3rd_cert:
      absent: true   # (or present: false, both makes sense to me, have no preference)
    name_of_4th_cert:
      absent: true
```

which is clear too, don't you think?

Anyway, these are just thoughts, I'm not using this formula atm, so jumped in to share an idea. If you are planning to submit a PR, you're free to write it as you prefer  :smile:

**Comment by ze42** [comment]
I do like. having the `absent: true` part directly in the same list.

Makes it easier to pass certificates, but disable/enable a specific certificates, without risking to have certificates that can be added/removed in the same run.

Also having the code in a single makes it easier to mange changes like if the actual filename changes, or how the path is taken/default/... better to have that part of code only once.
